Improve client recording of events such that clients are

a-robinson opened the pull request:
(1) less likely to drop events if the master is unavailable
(2) less likely to have goroutines block while trying to record an
event.

Done as part of #3163 to ensure that minions operate well even while the
master is down.

It feels a bit hacky to have a channel after the broadcast channel, but I didn't want to mess with the other uses of the broadcast type and this doesn't add much complexity.

If you feel that we should prefer dropping old events than new ones, just let me know. The only issue with that is that the working thread may continue trying to record an event for a couple minutes before giving up on it. That can be solved, but only by adding more complexity.

a-robinson committed:
Improve client recording of events such that clients are
(1) less likely to drop events if the master is unavailable
(2) less likely to have goroutines block while trying to record an
event.

Done as part of #3163 to ensure that minions operate well even while the
master is down.

a-robinson commented:
@lavalamp

lavalamp commented:
Looking-- already needs rebase though :)

lavalamp commented on pkg/client/record/event.go:
```diff
@@ -42,49 +55,82 @@ type EventRecorder interface {
 // or used to stop recording, if desired.
 func StartRecording(recorder EventRecorder, source api.EventSource) watch.Interface {
+	// Set up our own personal buffer of events so that we can clear out GetEvents'
+	// broadcast channel as quickly as possible to avoid causing the relatively more
```
If you're concerned about this, I think this is the wrong place to fix it, I think you should fix it at the other end of the queue, where Event() stuffs events in.

The way you have it, anyone who calls GetEvents could theoretically block events.

It's also a bit weird to have two 1000 element queues.

a-robinson replied:
The alternatives are to either get rid of the broadcast functionality that allows multiple subscribers or to mess with the watch.Broadcast class to allow for dropping rather than blocking. I'll do the latter during my free block of time later this afternoon unless I hear otherwise from you before then.

lavalamp replied:
That sounds fine-- perhaps you can add a ActionIfNotFull method to broadcaster. Thanks!

lavalamp commented on pkg/client/record/event.go:
```diff
@@ -42,49 +55,82 @@ type EventRecorder interface {
-			time.Sleep(1 * time.Second)
+		// Drop new events rather than old ones because the old ones may contain
+		// some information explaining why everything is so backed up.
+		if len(eventQueue) == maxQueuedEvents {
```
This isn't the way to do this with go, there's a race. I'm surprised go even lets you call len() on a channel.

```go
select {
case eventQueue <- event:
  //great
default:
  // it was full
}
```

a-robinson replied:
Wow, thanks for the code sample. I didn't know you could select based on adding an event to a channel.

I really wasn't too concerned about the race since the only race I see is dropping an event if the channel dropped from full to having a space in between the len() call and the dropping of the event, which should be very uncommon, but this is much nicer.

lavalamp commented on pkg/client/record/event.go:
```diff
@@ -25,11 +26,23 @@ import (
+var (
+	minSleep   = float64(1 * time.Second)
+	maxSleep   = float64(15 * time.Second)
+	backoffExp = 1.5
```
I don't think exponential backoff is great here, it's too hard to compute "how long must master be down before it starts losing events".

I'd suggest 1 try after a second, followed by N tries 15 seconds apart, or something of that nature.

lavalamp replied:
If you're worried about all the kubelets syncing up, maybe wait a random amount once.

erictune replied:
FYI util/wait has wait.Jitter()
On Jan 9, 2015 5:43 PM, "Daniel Smith" <notifications@github.com> wrote:

> In pkg/client/record/event.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3381#discussion-diff-22754771>
> :
>
> >  	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
> >
> >  	"github.com/golang/glog"
> >  )
> >
> > +const (
> > +	maxQueuedEvents  = 1000
> > +	maxTriesPerEvent = 10
> > +)
> > +
> > +var (
> > +	minSleep   = float64(1 * time.Second)
> > +	maxSleep   = float64(15 * time.Second)
> > +	backoffExp = 1.5
>
> If you're worried about all the kubelets syncing up, maybe wait a random
> amount once.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3381/files#r22754771>
> .
>

a-robinson replied:
Is it really an issue that it isn't easy to compute how long it'll be before an event gets dropped? Exponential backoff is common because it provides good real world behavior, not because it's easy to compute that total wait time.

For what it's worth, the total here is on the order of 2 minutes. but wait.Jitter adds a lot of variability anyways, since it adds an additional random fraction of the input wait time onto its result.

lavalamp replied:
Yeah, I'd like very easy-to-understand, predictable behavior here.

Suppose a master goes down for some period, and then comes back up. I'll posit that we should be able to reason about how many events the kubelets will have stored up, and what period of time they will deliver them to the master, and in what manner they will be delivered.

If the rule is "one wait randomly between 0 and N seconds, followed by M N second waits", then I think that when master comes back up, events ought to be delivered to it uniformly over N seconds.

Other rules, like this exponential backoff rule, don't let me predict the distribution or time period, at least not easily. I'm not even sure about the behavior of "N waits of rand(0, 5) seconds"-- I think, but am not sure, that this might cause event requests to be normally distributed instead of uniformly distributed, which would cause a much larger (though shorter) QPS spike.

Even if the above is overly paranoid, I'd still like the aggregate behavior of many clients releasing queued events to be very, very easy to reason about.

Exponential backoff is great for timing network packets to eventually miss each other. I don't actually think it's good for much else.

erictune replied:
@lavalamp reasoning seems good.

a-robinson replied:
The logic in the code you commented on is only governing the behavior while the master is still down. It would allow for a surge of requests when the master comes back up if the master isn't aggressive in returning errors, because this code only sleeps after a request fails. A success immediately resets the sleep to 0, which is potentially dangerous since each client is likely to have more than one event queued up.

I've implemented what you suggested, but it may be worth adding additional logic to slowly ease back onto the pedal once requests start succeeding again. Something along the lines of having an index into list of sleep times (or a counter + exponential backoff), and incrementing the index/counter foward or backward depending on whether each request succeeds.

a-robinson committed:
Push the full channel logic into the implementation of the broadcaster
in watch/mux.go rather than being in the client event recording code.

a-robinson commented:
Addressed your comments. Thanks for taking a look!

a-robinson committed:
Switch the client event recorder from exponential backoff to one random
sleep on the first failed request followed by a constant amount on all
subsequent consecutive failed requests.

lavalamp commented:
You've got a build failure :)

Will look later today

(I think ultimately making batch creation for events will be best)

lavalamp commented on pkg/client/record/event.go:
```diff
@@ -30,8 +31,9 @@ import (
-var retryEventSleep = 1 * time.Second
+const maxTriesPerEvent = 12
+
+var sleepDuration = time.Duration(10 * time.Second)
```
nit: time.Duration cast unneeded, since time.Second is a Duration.

lavalamp commented on pkg/client/record/event.go:
```diff
@@ -52,43 +54,59 @@ func StartRecording(recorder EventRecorder, source api.EventSource) watch.Interf
+			// Randomize the first sleep so that various clients won't all be
+			// synced up if the master goes down.
+			if tries == 1 {
+				time.Sleep(time.Duration(float64(sleepDuration) * rand.Float64()))
```
I am nervous because rand is not threadsafe...

a-robinson replied:
Wow, why isn't that mentioned in the doc page at all? Thanks for pointing it out.

lavalamp replied:
Sorry! I just realized that this was totally not true. The global one IS locked - http://golang.org/src/math/rand/rand.go#L160

Oops.

lavalamp commented on pkg/watch/mux.go:
```diff
@@ -22,6 +22,15 @@ import (
+
+const (
+	WaitIfChannelFull = iota
+	DropIfChannelFull = iota
```
Please type these, also iota means you don't have to say anything on the second line.

```
	WaitIfChannelFull FullChannelBehavior = iota
	DropIfChannelFull
```

lavalamp commented on pkg/watch/mux.go:
```diff
@@ -31,17 +40,27 @@ type Broadcaster struct {
-		watchers: map[int64]*broadcasterWatcher{},
-		incoming: make(chan Event, queueLength),
+		watchers:            map[int64]*broadcasterWatcher{},
+		incoming:            make(chan Event),
```
suggest still buffering this by a few so that it won't block even if Action is called a few times in rapid succession.

lavalamp commented on pkg/watch/mux_test.go:
```diff
@@ -95,7 +95,7 @@ func TestBroadcasterWatcherClose(t *testing.T) {
 func TestBroadcasterWatcherStopDeadlock(t *testing.T) {
 	done := make(chan bool)
-	m := NewBroadcaster(0)
+	m := NewBroadcaster(0, WaitIfChannelFull)
```
maybe a unit test here for DropIfChannelFull would be nice?

lavalamp commented:
Looking good-- only minor comments this time around :)

a-robinson committed:
Add a unit test for watch.Broadcaster DropIfChannelFull and a couple small fixes

a-robinson commented:
Incorporated all suggestions. Thanks Daniel!

a-robinson commented:
Ping - anything left that should be changed?

lavalamp commented:
LGTM, thanks
